lighthouse-core/audits/non-composited-animations.js

```diff
@@ -0,0 +1,178 @@
+  ACTIONABLE_FAILURE_REASONS.forEach(reason => {
+    if (failureCode & reason.flag) {
+      failureReasons.push(reason.text);
+      failureCode &= ~reason.flag;
+    }
+  });
+  return {failureReasons, hasNonActionable: Boolean(failureCode)};
+}
+
+class NonCompositedAnimations extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'non-composited-animations',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      title: str_(UIStrings.title),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['traces', 'AnimatedElements'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const traceOfTab = await TraceOfTab.request(trace, context);
+
+    /** @type {Map<string, {begin: LH.TraceEvent | undefined, status: LH.TraceEvent | undefined}>} */
+    const animationPairs = new Map();
+    traceOfTab.mainThreadEvents.forEach(event => {
+      if (event.name !== 'Animation') return;
+
+      if (!event.id2 || !event.id2.local) return;
+      const local = event.id2.local;
+
+      if (!animationPairs.has(local)) {
+        animationPairs.set(local, {begin: undefined, status: undefined});
+      }
+
+      const pair = animationPairs.get(local);
+      if (!pair) return;
+      switch (event.ph) {
+        case 'b':
+          pair.begin = event;
+          break;
+        case 'n':
+          pair.status = event;
+          break;
+      }
+    });
+
+    /** @type Map<string, {failureReasons: string[], nodes: LH.Audit.Details.NodeValue[]}> */
+    const animations = new Map();
+    animationPairs.forEach(pair => {
+      if (!pair.begin ||
+          !pair.begin.args.data ||
+          !pair.begin.args.data.nodeId ||
+          !pair.status ||
+          !pair.status.args.data ||
+          !pair.status.args.data.compositeFailed) return;
+
+      const nodeId = pair.begin.args.data.nodeId;
+      const element = artifacts.AnimatedElements.find(e => e.nodeId === nodeId);
+      if (!element) return;
+      /** @type LH.Audit.Details.NodeValue */
+      const node = {
+        type: 'node',
+        path: element.devtoolsNodePath,
+        selector: element.selector,
+        nodeLabel: element.nodeLabel,
+        snippet: element.snippet,
+      };
+
+      // Report animation only if all failure reasons are actionable
+      const {compositeFailed} = pair.status.args.data;
+      const {failureReasons, hasNonActionable} = getActionableFailureReasons(compositeFailed);
+      if (failureReasons.length === 0 || hasNonActionable) return;
+
+      const animation = '~placeholder~';
+      const data = animations.get(animation);
+      if (data) {
+        data.nodes.push(node);
+      } else {
+        // TODO: Add node specific failure reasons
```

Comment:
is this a TODO to resolve before merging or post merging?

Reply:
Before merging.